File: hybridsort.py
import threading
import time
from typing import List

SEGMENT_SIZE = 1000
THREAD_POOL_SIZE = 4

def custom_sort(arr: List[int], start: int, end: int) -> None:
    sorted = False
    while not sorted:
        sorted = True
        for i in range(start, end - 1):
            if arr[i] > arr[i + 1]:
                arr[i], arr[i + 1] = arr[i + 1], arr[i]
                sorted = False

def quick_sort(arr: List[int], low: int, high: int) -> None:
    if low < high:
        pi = partition(arr, low, high)
        quick_sort(arr, low, pi[0] - 1)
        quick_sort(arr, pi[1] + 1, high)

def partition(arr: List[int], low: int, high: int) -> List[int]:
    mid = low + (high - low) // 2
    if arr[low] > arr[mid]:
        arr[low], arr[mid] = arr[mid], arr[low]
    if arr[low] > arr[high]:
        arr[low], arr[high] = arr[high], arr[low]
    if arr[mid] > arr[high]:
        arr[mid], arr[high] = arr[high], arr[mid]

    pivot = arr[mid]
    i = low - 1
    j = low
    k = high
    while j <= k:
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
            j += 1
        elif arr[j] > pivot:
            arr[j], arr[k] = arr[k], arr[j]
            k -= 1
        else:
            j += 1
    return [i + 1, k]

def display_sorted_array(arr: List[int], size: int) -> None:
    print("Sorted Array: ", arr[:size])

def read_array_from_file(file_path: str) -> List[int]:
    with open(file_path, 'r') as file:
        content = file.read()
        number_strings = content.replace("\n", "").split(",")
        numbers = [int(num_str) for num_str in number_strings]
        return numbers

def measure_execution_time_and_memory(runnable) -> None:
    start_time = time.time_ns()
    runnable()
    end_time = time.time_ns()
    print("Execution Time: ", end_time - start_time, "nanoseconds")

def main():
    print()
    file_path = "listofnumbers.txt"
    array_to_sort = read_array_from_file(file_path)

    if array_to_sort is not None:
        executor = threading.Thread(target=measure_execution_time_and_memory, args=(lambda: quick_sort(array_to_sort, 0, len(array_to_sort) - 1),))
        executor.start()
        display_sorted_array(array_to_sort, SEGMENT_SIZE)

        executor = threading.Thread(target=measure_execution_time_and_memory, args=(lambda: custom_sort(array_to_sort, 0, len(array_to_sort) - 1),))
        executor.start()
        display_sorted_array(array_to_sort, len(array_to_sort))

    else:
        print("Error reading the input file.")
    print()

if __name__ == "__main__":
    main()

File: hybridsort.cc
#include <iostream>
#include <vector>
#include <fstream>
#include <sstream>
#include <thread>
#include <chrono>

const int SEGMENT_SIZE = 1000;

void customSort(std::vector<int>& arr, int start, int end) {
    bool sorted = false;
    while (!sorted) {
        sorted = true;
        for (int i = start; i < end - 1; i++) {
            if (arr[i] > arr[i + 1]) {
                std::swap(arr[i], arr[i + 1]);
                sorted = false;
            }
        }
    }
}

void quickSort(std::vector<int>& arr, int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

int partition(std::vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = (low - 1);
    for (int j = low; j <= high - 1; j++) {
        if (arr[j] < pivot) {
            i++;
            std::swap(arr[i], arr[j]);
        }
    }
    std::swap(arr[i + 1], arr[high]);
    return (i + 1);
}

void displaySortedArray(const std::vector<int>& arr, int size) {
    std::cout << "Sorted Array: ";
    for (int i = 0; i < size; i++) {
        std::cout << arr[i] << " ";
    }
    std::cout << std::endl;
}

std::vector<int> readArrayFromFile(const std::string& filePath) {
    std::vector<int> numbers;
    std::ifstream file(filePath);
    std::string str;
    while (std::getline(file, str, ',')) {
        numbers.push_back(std::stoi(str));
    }
    return numbers;
}

void measureExecutionTimeAndMemory(std::function<void()> func) {
    auto start = std::chrono::high_resolution_clock::now();
    func();
    auto end = std::chrono::high_resolution_clock::now();
    std::cout << "Execution Time: " << std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count() << " nanoseconds" << std::endl;
}

int main() {
    std::cout << std::endl;
    std::string filePath = "listofnumbers.txt";
    std::vector<int> arrayToSort = readArrayFromFile(filePath);

    if (!arrayToSort.empty()) {
        std::thread t1(& {
            measureExecutionTimeAndMemory(& { quickSort(arrayToSort, 0, arrayToSort.size() - 1); });
            displaySortedArray(arrayToSort, SEGMENT_SIZE);
        });
        t1.join();

        std::thread t2(& {
            measureExecutionTimeAndMemory(& { customSort(arrayToSort, 0, arrayToSort.size() - 1); });
            displaySortedArray(arrayToSort, arrayToSort.size());
        });
        t2.join();
    } else {
        std::cout << "Error reading the input file." << std::endl;
    }
    std::cout << std::endl;

    return 0;
}
